src/array.jl

```diff
@@ -327,3 +327,60 @@ function Base.reverse(v::CuVector, start=1, stop=length(v))
     stop < length(v) && copyto!(v′, stop+1, v, stop+1)
     return v′
 end
+
+function Base.reverse!(input::CuArray{T, N}, output::CuArray{T, N}; dims::Integer) where {T, N}
+
+    if !(1 ≤ dims ≤ length(size(input)))
+      ArgumentError("dimension $dims is not 1 ≤ $dims ≤ $length(size(input))")
+    end
+
+    if !all(size(input) .== size(output))
+      DimensionMismatch("input and output arrays are not of the same dimensions.")
+    end
+
+    nthreads = 256
+    nblocks = ceil(Int, length(input)/nthreads)
+    shmem = nthreads * sizeof(T)
+
+    shape = [size(input)...]
+
+    numelemsinprevdims = prod(shape[1:dims-1])
+    numelemsincurrdim = shape[dims]
+
+    function kernel(input::CuDeviceArray{T, N}, output::CuDeviceArray{T, N}) where {T, N}
```

Comment:
Why is this using shared memory?

If I simplify this a bit it looks like you are doing:
```
shared[idx] = in[I]
J = #= Index transformation =#
out[J] = shared[idx]
```

Reply:
CUDA beginner here. So, can I do this without using shared memory?

Reply:
Yeah, you only need shared memory when you want to communicate between threads. Here you just fetch some data on one thread, calculate the index to where it goes, and then push it.

You can do this (and it maybe be faster) but that would mean you are doing a index transform on the shared data and each thread is writing back linearly.

So instead of doing:
|tid |1 | 2 | 3 | 4|
| --- | - | ---| -- |--- |
|in[I] | 11 | 12 | 13 | 14 |
|shared | 11 | 12 | 13 | 14 |
|shared | 11 | 12 | 13 | 14 |
|out[J] |11 | 12 | 13 | 14 |

You might do

|tid |1 | 2 | 3 | 4|
| --- | - | ---| -- |--- |
|in[I] | 11 | 12 | 13 | 14 |
|shared | 11 | 12 | 13 | 14 |
|reg | 14 | 13 | 12 | 11 |
|out[J] | 14 | 13 | 12 | 11 |

which might improve the coalescing of the memory operations for the write.

Reply:
@vchuravy I decided to drop the idea of using shared memory all together. No point in adding complexity if it can be avoided. Did see significant speed up...

```
julia> @btime rev_d = my_reverse(A_d, dims=2)
  20.762 μs (60 allocations: 1.95 KiB)
```